Model/SubscriptionManagement.php

```diff
@@ -0,0 +1,113 @@
+
+use Magento\Customer\Model\Session;
+use Magento\Framework\App\RequestInterface;
+use Magento\Framework\Exception\LocalizedException;
+use Magento\Sales\Api\OrderManagementInterface;
+use Magento\Sales\Api\OrderRepositoryInterface;
+use Magento\Sales\Model\Order;
+use Paytrail\PaymentService\Api\SubscriptionLinkRepositoryInterface;
+use Paytrail\PaymentService\Api\SubscriptionRepositoryInterface;
+use Psr\Log\LoggerInterface;
+
+class SubscriptionManagement
+{
+    protected const STATUS_CLOSED = 'closed';
+    protected const ORDER_PENDING_STATUS = 'pending';
+
+    /**
+     * @var Session
+     */
+    protected $customerSession;
+
+    /**
+     * @var SubscriptionRepositoryInterface
+     */
+    protected $subscriptionRepository;
+
+    /**
+     * @var SubscriptionLinkRepositoryInterface
+     */
+    protected $subscriptionLinkRepository;
+
+    /**
+     * @var OrderRepositoryInterface
+     */
+    protected $orderRepository;
+
+    /**
+     * @var OrderManagementInterface
+     */
+    protected $orderManagementInterface;
+
+    /**
+     * @var RequestInterface
+     */
+    protected $request;
+
+    /**
+     * @var LoggerInterface
+     */
+    protected $logger;
+
+    /**
+     * @param Session $customerSession
+     * @param SubscriptionRepositoryInterface $subscriptionRepository
+     * @param SubscriptionLinkRepositoryInterface $subscriptionLinkRepository
+     * @param OrderRepositoryInterface $orderRepository
+     * @param OrderManagementInterface $orderManagementInterface
+     * @param RequestInterface $request
+     * @param LoggerInterface $logger
+     */
+    public function __construct(
+        Session $customerSession,
+        SubscriptionRepositoryInterface $subscriptionRepository,
+        SubscriptionLinkRepositoryInterface $subscriptionLinkRepository,
+        OrderRepositoryInterface $orderRepository,
+        OrderManagementInterface $orderManagementInterface,
+        RequestInterface $request,
+        LoggerInterface $logger
+    ) {
+        $this->customerSession = $customerSession;
+        $this->subscriptionRepository = $subscriptionRepository;
+        $this->subscriptionLinkRepository = $subscriptionLinkRepository;
+        $this->orderRepository = $orderRepository;
+        $this->orderManagementInterface = $orderManagementInterface;
+        $this->request = $request;
+        $this->logger = $logger;
+    }
+
+    /**
+     * @param $subscriptionId
+     * @return \Magento\Framework\Phrase
+     * @throws LocalizedException
+     */
+    public function cancelSubscription($subscriptionId)
+    {
+        try {
+            $subscription = $this->subscriptionRepository->get((int)$subscriptionId);
+            $orderIds = $this->subscriptionLinkRepository->getOrderIdsBySubscriptionId((int)$subscriptionId);
+            $customer = $this->customerSession->getCustomer();
+
+            foreach ($orderIds as $orderId) {
+                $order = $this->orderRepository->get($orderId);
```

Comment:
Performance warning repository load inside a loop. Prefer the usage of collection or getList methods over repository get when you are loading multiple entities.